### ISSUE TITLE
Several versions of AFWall+ are not starting during boot

### ISSUE THREAD

**Issue by 0pLuS0** [ukanth/afwall#901]
```System Specs:```
OnePlus 5T 
OxygenOS 5.1.7 - Oreo 8.1.0

AFWall+ 3.0.0
AFWall+ 3.0.1
AFWall+ 2.9.9

```There is something seriously wrong here with AFWall+ and the latest version of OxygenOS 5.1.7, I did not have this problem at all with OxygenOS 5.1.5```

When I restart the phone, I do not see the overhead notification saying the firewall is enabled, I'm assuming this is suppose to always be displayed when the firewall is running? Because if I go to the App Drawer and tap on the icon the notification then appears in the overhead.

I'm only using the ```Allow selected``` and I did not have ```Google Play Store``` allowed, and I connected online with the Play Store and I was able to download an app.

This is very disturbing that AFWall+ will not start on a system start, and that I have to manually start it.

I've attached two screen shots, one the overhead notification, again, I should always see this when the firewall is running? 

Also a screen shot of the notifications, which I could of sworn had 3 options under the ```Categories``` section when I was running OxygenOS 5.1.5, now I only see 2. So is a notification category missing?

```I'm attaching a logcat``` from 3.0.1, I hope you can figure out why AFWall+ is not starting...

I've also attached the RULES from 3.0.1, but this is when I start AFWall+ after rebooting, not sure how to get the rules without starting.

![1](https://user-images.githubusercontent.com/36190223/48750849-39387780-ec79-11e8-9c13-9fcb995e2855.jpg)

![2](https://user-images.githubusercontent.com/36190223/48750870-4bb2b100-ec79-11e8-961e-299a8b5c2338.jpg)

```LOGCAT```
[logcat.txt](https://github.com/ukanth/afwall/files/2598025/logcat.txt)

```RULES```
[IPv4rules.log](https://github.com/ukanth/afwall/files/2598181/IPv4rules.log)

**Comment by 0pLuS0** [comment]
I found an image I had for OxygenOS running 3.0.1 and it shows this ```3rd Notification - Boot Service``` that is not appearing in OxygenOS 5.1.7...

![boot_service](https://user-images.githubusercontent.com/36190223/48751206-faa3bc80-ec7a-11e8-848f-f91c1726b3b3.jpg)

**Comment by 0pLuS0** [comment]
I found an image I had for OxygenOS 5.1.5 running AFwall+ 3.0.1 and it shows this ```3rd Notification - Boot Service``` that is not appearing in OxygenOS 5.1.7...

![boot_service](https://user-images.githubusercontent.com/36190223/48751206-faa3bc80-ec7a-11e8-848f-f91c1726b3b3.jpg)

**Comment by 0pLuS0** [comment]
I noticed something else that seems odd, if you don't go into the ```UI Preferences``` then the ```Rules Progress``` Notification does not appear in the phone Settings > Notification for AFWall+ like in the above screen shots... 

hmm

**Comment by 0pLuS0** [comment]
I noticed something else that seems odd, if I don't go into the ```UI Preferences``` then the ```Rules Progress``` Notification does not appear in the phone Settings > Notification for AFWall+ like in the above screen shots... 

hmm

**Comment by ukanth** [comment]
Disable Xposed module. Xposed is making AFWall+ crashing and it's not able to start boot service since it died.

**Comment by 0pLuS0** [comment]
@ukanth sorry I didn't mention I had Xposed installed!

I originally had the module enabled, then I disabled it on all versions and the same problem, AFwall+ is not starting on a system startup.

Then I went back and reinstalled a twrp backup I had of 5.1.7 before I installed AFWall+ and I installed it, this time not enabling Xposed module and same problem, still not starting.

SORRY I didn't mention to you before, but I did all this already... 🤕

**Comment by ukanth** [comment]
Please provide logcat when you have xposed module disabled. logcat provided has the module active.

**Comment by 0pLuS0** [comment]
@ukanth 

Ok logcat with module disabled...

[logcat.txt](https://github.com/ukanth/afwall/files/2611948/logcat.txt)


Thanks

**Comment by ukanth** [comment]
Strange. There is no log related to afwall during boot. Is it full logcat of boot ?

**Comment by 0pLuS0** [comment]
@ukanth I cleared logcat then I restarted the phone, then took logcat....

I'll not clear it this time, and reboot the phone a few times, then take another one.

**Comment by 0pLuS0** [comment]
@ukanth 

I wiped the phone, then installed a TWRP backup that doesn't have AFWall+ installed. Then I installed AFwall+ from F-Droid, setup my rules and preferences, then rebooted, then took logcat again...

[logcat.txt](https://github.com/ukanth/afwall/files/2611989/logcat.txt)


Thanks

**Comment by ukanth** [comment]
Hmm.. it's weird. Did you disable bootservice using any other program (appops etc.,). There is no trace of afwall.

**Comment by 0pLuS0** [comment]
@ukanth I did not touch anything.

That is why I listed this as a critical bug, no version works for some reason on 5.1.7.

All of these steps I did exactly the same way in OxygenOS 5.1.5 and AFWall+ runs at startup without any problems.

So I can only assume OxygenOS did something to the ROM, or something I've done in these steps, even though the worked in 5.1.5 aren't working now in 5.1.7, so I will go back and wipe and install 5.1.7 Stock then only root it with Magisk and install AFWall+ and see if there's any difference.

These are the steps I did on OxygenOS.

1. I did a Deodex of only the services.jar this VDEX method;
https://gitlab.com/Nanolx/NanoDroid/blob/master/doc/DeodexServices.md

2. Next I used NanoDroid-patcher to apply signature spoofing.

3. Next I installed XposedInstaller_3.1.5-Magisk.apk

4. Next I installed Magisk

5. Next, In Magisk I installed these modules; Xposed Framework (SDK 27) - Debloater (Terminal Emulator) -  F-Droid Privileged Extension

6. Next I installed F-Droid and Termux from F-Droid

7. Next I Debloated Gapps/Google all of these;

Android Setup
Calendar
Chrome
Drive
Duo
Gboard
Gmail
Google
Google Backup Transport
Google One Time Init
Google Partner Setup
Google Pay
Google Play Movies
Google Play Music
Google Play Services
Google Services Framework
Market Feedback Agent
Maps
Photos
BugReportLite
EngSpecialTest
FactoryMode
Insight Provider
LogKitSdService
OnePlusLogKit
Oneplus System Service
RfToolkit

8. Next I made sure com.google.android.gms was deleted;
cd /data/data
rm -rf com.google.android.gms

9. Next I installed microG from F-Droid, these packages;
microG Services Core
microG Services Framework Proxy

After then installing and setting up microG I then installed these apps;

AnySoftKeyboard
Aurora Store
FX File Manager
```AFWall+```

10. I then setup the rules using ```Allow Selected``` picking a few applications;

Aurora Store - wifi
F-Droid - wifi
Magisk -wifi
Mozilla UnifiedNlP Backend - wifi
Nominatim Geocoder Backend - wifi
OnePlus Weather -wifi
Xposed -wifi

11. All AFWall+ preferences were left on default settings! The only thing I changed was ```Turn on log service```

12. I then tapped Enable Firewall and then Apply, and then I rebooted the phone.

**Comment by 0pLuS0** [comment]
Ok I went back and wiped all the partitions in TWRP and then flashed OxygenOS 5.1.7 you can see here the firmware;

https://www.oneplus.com/ca_en/support/softwareupgrade/details?code=7

I then installed only Magisk and AFwall+.

Next I only used ```Allow Selected``` and picked only Magisk for wifi, then I went into the UI Preferences and left everything on default and only enabled the Logging; ```Turn on log Service```, then I rebooted the phone and the same problem, AFWall+ does not start.

So it was nothing to do with all my steps of Deodex, or Debloating, and even Xposed is not installed on this ROM.

Also something I noticed is that the ```Rules Progress``` did not show up in the Notifications, after I rebooted I went back into the UI Preferences a second time, and now it appeared, odd...


Here's screen shots showing the Notifications, this time missing ```Rules Progress``` as I mentioned, and then App Drawer on Stock OxygenOS only Magisk & AFWall+


![app drawer](https://user-images.githubusercontent.com/36190223/48985859-95c9e580-f104-11e8-98f5-2c3010697895.jpg)



![notifications](https://user-images.githubusercontent.com/36190223/48985862-9d898a00-f104-11e8-8849-80305db9c1f0.jpg)

**Comment by 0pLuS0** [comment]
@ukanth 

Ok I went back and wiped all the partitions in TWRP and then flashed OxygenOS 5.1.7 you can see here the firmware;

https://www.oneplus.com/ca_en/support/softwareupgrade/details?code=7

I then installed only Magisk and AFwall+.

Next I only used ```Allow Selected``` and picked only Magisk for wifi, then I went into the UI Preferences and left everything on default and only enabled the Logging; ```Turn on log Service```, then I rebooted the phone and the same problem, AFWall+ does not start.

So it was nothing to do with all my steps of Deodex, or Debloating, and even Xposed is not installed on this ROM.

Also something I noticed is that the ```Rules Progress``` did not show up in the Notifications after I rebooted. I had to go back into the UI Preferences a second time, and now the Notification setting appears, hmm odd...


Here's screen shots showing the Notifications, this time missing ```Rules Progress``` as I mentioned, and then App Drawer on Stock OxygenOS only Magisk & AFWall+

```Logcat attached at the bottom```

![app drawer](https://user-images.githubusercontent.com/36190223/48985859-95c9e580-f104-11e8-98f5-2c3010697895.jpg)



![notifications](https://user-images.githubusercontent.com/36190223/48985862-9d898a00-f104-11e8-8849-80305db9c1f0.jpg)


```HERE IS A LOGCAT```

[logcat.txt](https://github.com/ukanth/afwall/files/2613590/logcat.txt)

**Comment by 0pLuS0** [comment]
@ukanth I have finally been able to get AFWall+ to run several times at startup and show all the Notification options this time.

The important thing I want you to consider/realize is that the only thing you mentioned was Xposed causing an issue, but I only once had it enabled on numerous tests with it disabled and never getting AFWall+ to start. Beyond the Xposed module, how else could an end-user cause it to not start, I personally don't even see how that is possible, on a new install of a clean rom, everything default settings, and not doing anything also special in AFWall+, only enabling the logs...

What I have now found out after messing with this for a week and many installs, is that I've had to do things in a particular order to make this happen. I am not aware that AFWall+ has to be started and setup in any particular order/steps to make this work, this does seem like a bug that I have to do things in a certain order to make this work.

Here is what I've done to finally make this work.

1. Install and open AFWall+
2. Allow Selected marked - then I clicked on a few applications I wanted to allow
3.I clicked Enable Firewall
4.Preferences - UI Preferences - Unchecked & checked - Enable notifications & show notification icon
5. Log - Turn on log service
6. Main window - tapped "Apply"

Now I rebooted and AFWall+ started on system start.

Number 4 was the only thing different I have ever done in all the tests I have been working on. Unchecking and then checking these 2 notifications options seems to have made a difference, or else just some strange coincidence. But the thing I want you to understand is that I never did this in the past, and AFWall+ never started...

```I'm also attaching a Logcat, maybe between this one and the others you'll finally figure out what has been going on.```

[logcat.txt](https://github.com/ukanth/afwall/files/2617955/logcat.txt)


I personally feel like there's some notification bug going on, after all I started another issue in the past about the App Drawer notification icon that won't go away.

https://github.com/ukanth/afwall/issues/898

Here's a screen shot now showing all the Notifications working.

![notifications](https://user-images.githubusercontent.com/36190223/49052719-f92b4480-f1e4-11e8-812f-1ade30ab9de0.jpg)

**Comment by 0pLuS0** [comment]
@ukanth I have finally been able to get AFWall+ to run several times at startup and show all the Notification options this time.

The important thing I want you to consider/realize is that the only thing you mentioned was Xposed causing an issue, but I only once had it enabled on numerous tests, and with it disabled AFWall+  would never start. 

Beyond the Xposed module, how else could an end-user cause it to not start, I personally don't even see how that is possible. On a new install of a clean rom, everything default settings, and not doing anything special in AFWall+, I only enabled the log.

What I have now found out after messing with this for a week and many installs, is that I've had to do things in a particular order to make this happen. I am not aware that AFWall+ has to be started and setup in any particular order/steps to make this work, this does seem like a bug that I have to do things in a certain order to make this work.

Here is what I've done to finally make this work.

Install and open AFWall+
Allow Selected marked - then I clicked on a few applications I wanted to allow
I clicked Enable Firewall
Preferences - UI Preferences - Unchecked & checked - Enable notifications & show notification icon
Log - Turn on log service
Main window - tapped "Apply"

Now I rebooted and AFWall+ started on system start.

The only thing different I have ever done in all the tests I have been working on was, unchecking and then checking the Enable notifications & show notification icon options, which seems to have made a difference, or else just some strange coincidence. But the thing I want you to understand is that I never did this in the past, and AFWall+ never started...

```I'm also attaching a Logcat, maybe between this one and the others you'll finally figure out what has been going on.```

[logcat.txt](https://github.com/ukanth/afwall/files/2617955/logcat.txt)


I personally feel like there's some notification bug going on, after all I started another issue in the past about the App Drawer notification icon that won't go away.

https://github.com/ukanth/afwall/issues/898

Here's a screen shot now showing all the Notifications working.

![notifications](https://user-images.githubusercontent.com/36190223/49052719-f92b4480-f1e4-11e8-812f-1ade30ab9de0.jpg)

**Comment by 0pLuS0** [comment]
@ukanth I have finally been able to get AFWall+ to run several times at startup and show all the Notification options this time.

The important thing I want you to consider/realize is that the only thing you mentioned was Xposed causing an issue, but I only once had it enabled on numerous tests, and with it disabled AFWall+  would never start. 

Beyond the Xposed module, how else could an end-user cause it to not start, I personally don't even see how that is possible. On a new install of a clean rom, everything default settings, and not doing anything special in AFWall+, I only enabled the log.

What I have now found out after messing with this for a week and many installs, is that I've had to do things in a particular order to make this happen. I am not aware that AFWall+ has to be started and setup in any particular order/steps to make this work, this does seem like a bug that I have to do things in a certain order to make this work.

Here is what I've done to finally make this work.

Install and open AFWall+
Allow Selected marked - then I clicked on a few applications I wanted to allow
I clicked Enable Firewall
Preferences - UI Preferences - Unchecked & checked - Enable notifications & show notification icon
Log - Turn on log service
Main window - tapped "Apply"

Now I rebooted and AFWall+ started on system start.

The only thing different I have ever done in all the tests I have been working on, was unchecking and then checking the Enable notifications & show notification icon options, which seems to have made a difference, or else just some strange coincidence. But the thing I want you to understand is that I never did this in the past, and AFWall+ never started...

```I'm also attaching a Logcat, maybe between this one and the others you'll finally figure out what has been going on.```

[logcat.txt](https://github.com/ukanth/afwall/files/2617955/logcat.txt)


I personally feel like there's some notification bug going on, after all I started another issue in the past about the App Drawer notification icon that won't go away.

https://github.com/ukanth/afwall/issues/898

Here's a screen shot now showing all the Notifications working.

![notifications](https://user-images.githubusercontent.com/36190223/49052719-f92b4480-f1e4-11e8-812f-1ade30ab9de0.jpg)

**Comment by gashtal** [comment]
I am having the same problem. I recently purchased a new phone (ASUS Zenfone 6), but Afwall does not seem to start automatically when I reboot the phone. I am using the stuck ROM which is Android 9, with unlocked bootloader and rooted using magisk. No other modifications are done.

My observations:

Afwall does not start automatically when the phone is rebooted, and all traffic is allowed after reboot. Afwall's notification does not appear in status bar either, after reboot. I have tried disabling the battery optimization which is introduced in Android 9 for Afwall, adding a startup delay, and enabling the startup leak prevention option from experimental settings, none of which seem to have any positive effect.

After reboot, if Afwall is opened manually, then superuser permission acquired message is seen and Afwalls' persistent notification appears on the status bar and will start working. However, firewall rules are only applied if WiFi/Data is NOT enabled before opening Afwall; if they are, then Afwall will allow all network traffic. Firewall rules will only start working in this case if the WiFi/Data connection is disabled, and rules are reapplied in Afwall, and then the network connection is enabled. Albeit, this doesn't seem to always work, sometimes the rules need to be reapplied multiple times until they actually get applied (maybe related to https://github.com/ukanth/afwall/issues/1031?). A similar thing also happens frequently when rules are changed in Afwall and then reapplied, in which case Afwall will suddenly stop applying all the rules and allows all internet traffic until the rules are reapplied multiple times and eventually stick.

Unfortunately these problems are making Afwall very hard to use on my new phone and I have to be alert all the time and use random apps that are not supposed to access internet frequently to make sure Afwall is actually working. I used this firewall on my older phone with LineageOS 14 for years, and never had such issues. Not sure if the problem is from the specific ROM on my new phone, some incompatibility between Afwall and Android 9, or maybe even magisk. On a side note, I actually went ahead and also tried NoRoot Firewall and NetGuard; both seemed to fail to work at startup like Afwall.

**Comment by steffenl82** [comment]
Is there any chance that u are using a sdcard configured as "adoptable storage"? If afwall is installed on the sdcard I observe the same problems (autostart not working, and boot service notification missing). However, once I move afwall to the internal storage both problems are gone

**Comment by gashtal** [comment]
@steffenl82 I don't think so. I never set the SD Card to be used as adoptable storage and I am not sure if ASUS's ROM even supports it since I cannot find any information/settings in the ROM about this. I can safely remove my SD Card, too, which I think means it is being used as a separate dedicated storage point.

**Comment by mikemabi** [comment]
I noticed when I rebooted, and afwall+ didn't start at boot, and I started it myself immediately upon reboot, that it got killed again...

Made it a system app (using a magisk module), and problem was solved. By the way there were other apps as well that didn't start at boot which were worked fine before upgrading to Oreo ( Android 8.1) Moved to system, and they also worked fine too! 

It seems to be a problem of the OS restricting some user apps from starting at boot

**Comment by ukanth** [comment]
No inputs. Closing due to inactivity. Please reopen if it's still relevant.
